### DISCUSSION TITLE
MvvmDialogs for Avalonia

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#7130] [General]
I have just refactored MvvmDialogs to support Avalonia

Code still needs testing and isn't ready to be published, but [beta is available from my fork here.](https://github.com/mysteryx93/mvvm-dialogs) (doc isn't updated yet)

Using IDialogService is a whole lot better than the [solution written in the Avalonia docs!](https://docs.avaloniaui.net/tutorials/music-store-app/opening-a-dialog)

**Comment by robloo** [comment]
Great work! Dialogs is certainly an area I would like to see improvement on in Avalonia.

**Comment by mysteryx93** [comment]
That library makes it sweet and simple. I can't understand that someone would write an application without it.

I still need to port the Demos to Avalonia. "Some" of the WPF demos (.NET Core) are working, others still need fixing to compile.

With a bit more work, it can support WPF, Avalonia, UWP, WinUI3 and Blazor -- any MVVM application needs that. From my part I'll just focus on Avalonia.

**Reply by robloo** [reply]
Will definitely take a look at using this sometime. Do you think it could be merged upstream as well with https://github.com/FantasticFiasco/mvvm-dialogs?

**Reply by mysteryx93** [reply]
It will be merged. [Here's the work-in-progress thread.](https://github.com/FantasticFiasco/mvvm-dialogs/issues/177) Fantastic Fiasco still has plenty of work to do on his side to test, polish and document the code before public release.

This release will have breaking changes that will need to be documented
- ShowDialog methods are now Async
- Framework dialog settings are now platform-agnostic
- Custom factories are structured differently
- Namespace changes

**Comment by mysteryx93** [comment]
I've created the first 2 demos. [Here's a sample ViewModel to show and close another window.](https://github.com/mysteryx93/mvvm-dialogs/blob/master/samples/Avalonia/Demo.CloseNonModalDialog/MainWindowViewModel.cs)

The part that's bogging me, however, is [class registration with ReactiveUI/Splat.](https://github.com/mysteryx93/mvvm-dialogs/blob/master/samples/Avalonia/Demo.CloseNonModalDialog/App.axaml.cs#L15) Could someone enlighten me on this one?

Instead of declaring the usual

    Locator.Register<IDialogService, DialogService>()

It's asking me to create the object myself with all constructor parameters. In practice, a ViewModel may have 5 dependencies, and each of them may have 3 to 6 dependencies -- some of which repeat. Am I expected to do this whole initialization tree manually? That's the whole point of Dependency Injection, to do that job for me isn't it? Any advice here would be appreciated.

**Comment by mysteryx93** [comment]
Added some cool new feature: the ViewModel can implement ICloseable and IActivable with events RequestClose and RequestActiate. Raising those events will automatically close or activate the View!

Several of the Avalonia demos are done and it's working beautifully.
